drt_1d.py

```diff
@@ -156,32 +252,113 @@ def perform_calculation(input_data:dict):
     - numpy.ndarray: The neutron flux vector calculated from the input data.
     """
     logging.debug(f"Performing calculation with input data: {input_data}")
-    x_min = input_data.get("minimum_x")
-    x_max = input_data.get("maximum_x")
-    num_voxels = input_data.get("number_of_voxels")
-    source_voxel_index = input_data.get("source_voxel_index")
-    sigma_s_values = input_data.get("scattering_cross_section")
-
-    voxel_size = (x_max - x_min) / num_voxels
+    
+    voxels_data = input_data['voxels']
+    if not isinstance(voxels_data, list) or len(voxels_data) == 0:
+        error_msg = "Voxels data must be a non-empty list."
+        logging.error(error_msg)
+        raise ValueError(error_msg)
+    
+    voxels:list[Voxel] = []
+    for voxel_data in voxels_data:
+        voxel = Voxel(
+            index=voxel_data["index"],
+            start_position=voxel_data["start_position"],
+            end_position=voxel_data["end_position"],
+            scattering_cross_section=voxel_data["scattering_cross_section"],
+            positions=voxel_data["positions"]
+        )
+        voxels.append(voxel)
+    if not voxels:
+        error_msg = "No voxels found in input data."
+        logging.error(error_msg)
+        raise ValueError(error_msg)
+    
+    source_voxel_index = input_data['source_voxel_index']
+    if not(0 <= source_voxel_index < len(voxels)):
+        error_msg = "Source voxel index out of bounds."
+        logging.error(error_msg)
+        raise ValueError(error_msg)
+    
+    sigma_s_values = [voxel.scattering_cross_section for voxel in voxels]
+    num_voxels = len(voxels)
     
     s = construct_source_term(num_voxels, source_voxel_index)
-    K_trans = construct_transition_matrix(num_voxels, voxel_size, sigma_s_values)
+    K_trans = construct_transition_matrix(voxels)
     phi = calculate_neutron_flux(K_trans, sigma_s_values, s)
     
-    logging.info(f"Scattering Cross-section Vector (sigma_s):\n{sigma_s_values}")
-    logging.info(f"Source Vector (s):\n{s}")
-    logging.info(f"Transition Matrix (K_trans):\n{K_trans}")
-    logging.info(f"Neutron Flux Vector (phi):\n{phi}")
-    
     return phi
 
+def validate_input(input_data:dict) -> None:
+    """
+    Validates the overall input data structure and key parameters.
+    
+    NOTE: possible implementation could look at start and end positions of the voxels and ensure the user did create gaps.
+
+    Parameters:
+    - input_data (dict): The input data to be validated.
+
+    Returns:
+    - None
+
+    Raises:
+    - ValueError: If any validation checks fail.
+    """
+    logging.debug("Validating input data")
+    try:
```

Comment:
I think an overarching `try` block like this is a big hammer.  You are doing a lot of careful checking and raising errors where they occur - you don't need a try block